### ISSUE TITLE
[API cleanup] flux_attr functions need a makeover

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#1818]
The functions `flux_attr_get()`, and `flux_attr_set()` wrap synchronous RPCs.

We may want to switch to something that doesn't hide blocking RPCs, and that makes the internal caching of attributes a bit more explicit so that an application can fetch the value of an attribute they know they cached earlier without needing to create a future to access the cache.

Also `flux_attr_fake()` (used only in test) has got to go.

**Comment by garlick** [comment]
Looking at this a bit more closely, I'm not sure it's a problem having `flux_attr_get()` and `flux_attr_set()` wrap synchronous RPCs, since when used in a module implementing a service, typically they run during initialization, before the service begins.  The cache allows commonly used values like `size` and `rank` to be reused later without an RPC.  Switching to a future-based interface would just require people to provide a bunch of extra code that isn't helping them.  Possibly documentation could be improved a bit so there are at least no surprises here.

The _flags_ could be cleaned up though:
```c
/* Flags can only be set by the broker.
 */
enum {
    FLUX_ATTRFLAG_IMMUTABLE = 1,    /* attribute is cacheable */
    FLUX_ATTRFLAG_READONLY = 2,     /* attribute cannot be written */
                                    /*   but may change on broker */
    FLUX_ATTRFLAG_ACTIVE = 4,       /* attribute has get and/or set callbacks */
};

const char *flux_attr_get (flux_t *h, const char *name, int *flags);
```
There are no users "in the wild" of this (optional) flags parameter to `flux_attr_get()`.  The only flag that's even useful on the client side is IMMUTABLE, which determines whether a value can be cached indefinitely.  Since the cache is not directly exposed, even that could be hidden from the API.  I think we should drop the flags definitions and the flags parameter from `flux_attr_get()`.

The `flux_attr_first()` and `flux_attr_next()` iterators are only used in the `flux lsattr` command, and I can't think of a reason these would be generally useful, so perhaps `flux lsattr` should just construct the RPC directly and we should remove these from the API.

Finally `flux_attr_fake()` is mainly used in for brokerless testing.  Possibly this could be made private to flux-core, and the `loop://` connector could put a `flux_opt_set()` interface in front of it for external users?

One thing that's prevented this API from getting cleaned up before is it's connected with the general concept of configuration, for which we have not developed a grand plan.  However, I think given that flux API use is likely to be growing, we should do the minor cleanups now and at least simplify it as much as possible to lessen the pain later if we need to change it.

**Comment by garlick** [comment]
In addition, probably the code in `libflux/info.c` should be folded into `libflux/attr.c` since `flux_get_rank()` and `flux_get_size()` are just wrappers around `flux_get_attr()`.

The `flux_get_nodeset()` function should be culled since it only has one user in `flux-module`.

**Comment by garlick** [comment]
Closed by #1845.
